feat: 사용자 기능 구현

MCshin00 opened the pull request:
### PR 타입
- [x] 기능 추가
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 반영 브랜치
- feature/user-api -> main

### 변경 사항
- jwt를 사용한 회원가입, 로그인 기능 및 회원 관련 CRUD 기능 구현
- 생성, 수정, 삭제 관련 정보를 담는 BaseEntity 구현

MCshin00 committed:
feat: 로그인 및 회원가입 기능 구현

build.gradle 의존성 정리
인증/인가는 filter에서 처리하도록 구현

MCshin00 committed:
refactor: Valid 유효성 검사 예외 처리 책임 이동

기존 Valid 유효성 검사는 Controller의 signup메서드에서 직접 처리했는데, 이를 GlobalExceptionHandler에서 처리하도록 수정
그 외 사용하지 않는 import 정리

MCshin00 committed:
feat: 회원정보 수정 기능 구현

일반 사용자와 관리자가 사용할 수 있는 회원정보 수정 API를 분리 구현
권한에 맞지 않는 요청을 할 경우 예외 메세지를 출력하도록 구현

MCshin00 committed:
feat: 데이터 생성, 수정, 삭제 정보 구현

MCshin00 committed:
refactor: 사용자 정보 수정 기능 응답 코드 변경

기존 201 CREATED에서 200 OK로 변경함

MCshin00 committed:
feat: 회원 탈퇴 기능 구현

삭제 기능은 soft-delete로 DB의 삭제 정보를 업데이트하도록 구현

MCshin00 committed:
feat: 회원정보조회 기능 구현

MCshin00 committed:
refactor: 로그인 및 회원가입에 대한 예외처리 추가

회원가입 및 로그인은 로그인 되지 않은 상태에서만 시도할 수 있도록 변경
로그인 요청이 POST 메서드가 아닌 경우 예외처리 추가

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/controller/UserController.java:
```diff
@@ -0,0 +1,89 @@
+        // 성공 응답으로 201 Created와 사용자 ID 반환
+        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto);
+    }
+
+    @GetMapping("/")
+    public ResponseEntity<?> getUserInfo(@RequestParam String username) {
+        // 사용자 정보 가져오기
+        UserResponseDto responseDto = userService.getUserInfo(username);
+
+        // 성공 응답으로 200 OK와 사용자 ID 반환
+        return ResponseEntity.status(HttpStatus.OK).body(responseDto);
+    }
+
+    @PutMapping("/")
+    public ResponseEntity<?> updateUser(@Valid @RequestBody UpdateUserRequestDto requestDto, @AuthenticationPrincipal UserDetailsImpl userDetails) {
+        // 사용자 정보 업데이트
+        UserIdResponseDto responseDto = userService.updateUser(requestDto, userDetails.getId());
+
+        // 성공 응답으로 200 OK와 사용자 ID 반환
+        return ResponseEntity.status(HttpStatus.OK).body(responseDto);
+    }
+
+    @Secured({"ROLE_MANAGER", "ROLE_MASTER"})
+    @PutMapping("/{userId}")
+    public ResponseEntity<?> updateUser(@Valid @RequestBody UpdateUserRequestDto requestDto, @PathVariable Long userId) {
+        // 사용자 정보 업데이트
+        UserIdResponseDto responseDto = userService.updateUser(requestDto, userId);
+
+        // 성공 응답으로 200 OK와 사용자 ID 반환
+        return ResponseEntity.status(HttpStatus.OK).body(responseDto);
+    }
+
+    @DeleteMapping("/")
+    public ResponseEntity<?> deleteUser(@AuthenticationPrincipal UserDetailsImpl userDetails) {
+        // 사용자 정보 삭제(soft-delete)
+        UserIdResponseDto responseDto = userService.deleteUser(userDetails.getId());
+
+        // 성공 응답으로 200 OK와 사용자 ID 반환
+        return ResponseEntity.status(HttpStatus.OK).body(responseDto);
+    }
```
ResponseEntity를 사용하여 status를 지정하는 방법도 직관적이고 좋지만
Controller 메서드에서 반환하는 response status는 대부분 HttpStatus.OK이기 때문에 
따로 ResponseDto 클래스를 만들어 HttpStatus.OK를 상태를 가진 응답을 반환하는 메서드를 구현하고 사용하는 것은 어떤지 제안합니다!

MCshin00 replied:
고민해보겠습니다!

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/entity/User.java:
```diff
@@ -0,0 +1,53 @@
+
+    private String imageUrl;
+
+    public User(String username, String password, String email, String phoneNumber, UserRoleEnum role) {
+        this.username = username;
+        this.password = password;
+        this.email = email;
+        this.phoneNumber = phoneNumber;
+        this.role = role;
+        this.publicProfile = true;
+    }
```
객체 필드 값을 대입할 때 @Builder 어노테이션을 사용하는 것보다 직접 값을 대입하는 형식으로 구현하셨는데 이유가 있을까요?

MCshin00 replied:
Builder 사용하는걸 까먹었네요... 따로 이유는 없었고, Builder 사용하는 방향으로 리팩토링 해보겠습니다.

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/service/UserService.java:
```diff
@@ -0,0 +1,106 @@
+    private final PasswordEncoder passwordEncoder;
+    private final AuditorAware<String> auditorAware;
+
+    public UserIdResponseDto signup(@Valid SignupRequestDto requestDto) {
+        String username = requestDto.getUsername();
+        String password = passwordEncoder.encode(requestDto.getPassword());
+        String email = requestDto.getEmail();
+        String phoneNumber = requestDto.getPhoneNumber();
+        UserRoleEnum role = requestDto.getRole();
+
+        checkUsername(username);
+        checkEmail(email);
+        checkPhoneNumber(phoneNumber);
+
+        User user = new User(username, password, email, phoneNumber, role);
+        user.setCreatedBy(username);
+        User savedUser = userRepository.save(user);  // User 엔티티 저장
+
+        return new UserIdResponseDto(savedUser.getId());  //저장된 User Entity의 id값을 통해 SignupResponseDto를 생성하고 반환
+    }
```
제가 코드를 이해했을 때 일반 회원과 관리자 회원의 회원가입 로직 차이가 없는 것으로 이해했습니다.
혹시 차이가 없는 것이 맞다면 저희 과제했던 것처럼 관리자 회원가입일 경우  관리자 전용 키 값을 입력받아 관리자 계정을 만들 수 있는 권한을 가진 사람인지 유효성 체크하는 로직이 필요하다고 생각됩니다.
어떻게 생각하시는지 궁금합니다!

MCshin00 replied:
관리자 회원가입을 분리하는 것은 생각하고 있었는데, 어떤 방식으로 구현하면 좋을지 고민이 돼서 적용을 못 했던 부분이었습니다. 과제에서 했던 것처럼 관리자 전용 키 값을 입력하는 방식은 강의에서 실제로 이렇게 허술하게 관리하지 않는다고 하셨던 걸로 기억하는데, 다른 로직이 필요하다고 생각합니다!

strongcookdas commented on blackWhiteDeliveryDriver/src/main/java/com/sparta/blackwhitedeliverydriver/service/UserService.java:
```diff
@@ -0,0 +1,106 @@
+    public UserResponseDto getUserInfo(String username) {
+        User user = userRepository.findByUsername(username)
+                .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다."));
+
+        return new UserResponseDto(user);
+    }
+
+    @Transactional
+    public UserIdResponseDto updateUser(@Valid UpdateUserRequestDto requestDto, Long userId) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다."));
+
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setEmail(requestDto.getEmail());
+        user.setPhoneNumber(requestDto.getPhoneNumber());
+        user.setImageUrl(requestDto.getImgUrl());
+        user.setPublicProfile(requestDto.isPublicProfile());
+        user.setRole(requestDto.getRole());
+
+        userRepository.save(user);
+
+        return new UserIdResponseDto(user.getId());
+    }
+
+    @Transactional
+    public UserIdResponseDto deleteUser(Long userId) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다."));
+
+        // 삭제를 수행한 사용자의 username을 가져옵니다.
+        String deletedBy = auditorAware.getCurrentAuditor()
+                .orElseThrow(() -> new IllegalStateException("로그인 정보 없음"));
+
+        user.setDeletedBy(deletedBy);
+        user.setDeletedDate(LocalDateTime.now());
+
+        userRepository.save(user);
+
+        return new UserIdResponseDto(user.getId());
+    }
```
EnumType을 통해서 예외 문구를 관리하는 것은 어떻게 생각하시나요??
반복적인 예외 문구를 관리할 수 있을 것 같아 제안합니다!

[참고 링크]
[예외 처리 관련 링크](https://goldenrabbit.co.kr/2024/04/03/spring-%EC%8A%A4%ED%94%84%EB%A7%81-%EB%B6%80%ED%8A%B8-%EC%98%88%EC%99%B8-%EC%B2%98%EB%A6%AC-%EA%B0%80%EC%9D%B4%EB%93%9C/)

MCshin00 replied:
예외 문구는 Enum으로 처리하도록 리팩토링해보겠습니다!

strongcookdas reviewed (commented):
코드를 너무 잘 구현하셔서 제가 많이 배웠습니다.
몇 개의 코멘트를 남겼습니다. 시간 있으실 때 확인 부탁드립니다!

MCshin00 committed:
Merge branch 'Black-White-Delivery-Driver:main' into feature/user-api

MCshin00 committed:
refactor: User 엔티티의 Id 필드 제거 및 Builder 추가

User의 PK를 Long 타입의 Id에서 username 필드로 변경
직접 구현한 생성자를 제거하고 Builder 어노테이션을 적용
userIdResponseDto는 username을 응답하도록 클래스명 및 필드 타입 변경

MCshin00 committed:
refactor: 관리자 전용 회원가입 API 추가

MCshin00 committed:
refactor: 예외 메세지 관리 Enum 추가
